### DISCUSSION TITLE
when will the __webpack_require__.hmd be defined ?

### DISCUSSION THREAD

**Discussion by lanmch** [webpack/webpack#14554] [General]
# Bug report

**What is the current behavior?**
![image](https://user-images.githubusercontent.com/37408756/138458499-c8d7d8b1-8fb3-4bf5-bae4-0f9b26bd92bf.png)
i get the error about hmd not defined

**If the current behavior is a bug, please provide the steps to reproduce.**
I think when I upgrade version 4 to version 5, the treeshaking might be better. But maybe in my project, there is some unused module, and it was tree shaked. But maybe these modules is used by Node service.
So, to be exactly, I mainly want to know when will the __webpack_require__.hmd will be defined

I am expecting a reply

**What is the expected behavior?**

**Other relevant information:**
webpack version:  5.51.1
Node.js version: 14.16
Operating System: macOs

**Comment by alexander-akait** [comment]
Please use `Discussions` for questions in future, thanks! Can you provide configuration? I think you have multiple HotModulesReplacement plugins

**Reply by lanmch** [reply]
HotModulesReplacement? the error is online of the production mode

**Reply by alexander-akait** [reply]
You should not have `hot` code in production

**Reply by alexander-akait** [reply]
Check your plugins array

**Reply by lanmch** [reply]
i'm sure in my plugins, there is 0 configuration about hot

**Reply by alexander-akait** [reply]
Unfortunately, you do not provide enough information to help you. It would be nice to see the configuration.

**Reply by alexander-akait** [reply]
hm, do you minimize project for production usage?

**Reply by lanmch** [reply]
> 嗯，您是否将项目最小化以用于生产用途？

like uglify or terser? yes i use it

**Reply by alexander-akait** [reply]
What is you use and how you use?

**Reply by lanmch** [reply]
the mode of production will open minimize? Also I append the terser plugin to minimizer like this
```javascript
if (shouldUglifyJS()) {
    config.optimization.minimizer.push(
      new TerserPlugin({
        extractComments: false,
      }),
    );
  }
```

**Reply by alexander-akait** [reply]
Can you try to disable it and run again and check errors

**Reply by lanmch** [reply]
i have tried it .
![image](https://user-images.githubusercontent.com/37408756/138465693-6183e656-263e-42bb-9301-651dcc624009.png)

**Reply by alexander-akait** [reply]
Looks like one of your plugins uses old webpack api, my recommendation - start to comment plugins step by stepa and check it

**Reply by lanmch** [reply]
i think it might be a big work, this project config has 900 lines😭

**Reply by lanmch** [reply]
so I want to know what will define the __webpack_require__.hmd

**Reply by alexander-akait** [reply]
Unfortunately, I don't see another solution. If you could let me to look at this, I think we would solve it much faster.

**Reply by lanmch** [reply]
webpack_require .hmd  , what will define it? or what configuration do you want to know , i can show u

**Reply by alexander-akait** [reply]
https://github.com/wallabyjs/public/issues/2651, maybe karma-webpack?

**Reply by alexander-akait** [reply]
Try to update all webpack dependencies before testing webpack v5

**Reply by lanmch** [reply]
is there any docs?

**Reply by alexander-akait** [reply]
About what? We don't have `webpack_require.hmd` in source code, it appears from plugin

**Reply by lanmch** [reply]
do you know what plugins  will include it? this is define code
```javascript
/* webpack/runtime/harmony module decorator */
/******/ 	!function() {
/******/ 		__webpack_require__.hmd = function(module) {
/******/ 			module = Object.create(module);
/******/ 			if (!module.children) module.children = [];
/******/ 			Object.defineProperty(module, 'exports', {
/******/ 				enumerable: true,
/******/ 				set: function() {
/******/ 					throw new Error('ES Modules may not assign module.exports or exports.*, Use ESM export syntax, instead: ' + module.id);
/******/ 				}
/******/ 			});
/******/ 			return module;
/******/ 		};
/******/ 	}();
```

**Reply by alexander-akait** [reply]
Sorry, no, try to search inside code of you plugins, also you can try to search in node_modules

**Reply by lanmch** [reply]
i find the file, 'node_modules/webpack/lib/runtimeGlobal' , this file show all property can be defined by webpack_require . will it has some influence？

**Reply by alexander-akait** [reply]
`webpack_require` is common object for runtime functions, bug no `hmd` function...
